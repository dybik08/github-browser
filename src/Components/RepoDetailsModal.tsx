import React, { useState } from 'react';
import { Button, Modal } from 'antd';
import { GitForkIcon, RepoIcon, StarIcon } from '@primer/octicons-react';
import githubLangColors from '../github-lang-colors';

import { Collapse } from 'antd';
import axios from 'axios';

const { Panel } = Collapse;

const RepoDetailsModal = (props: any) => {
    const [reposData, setReposData] = useState([]);

    const onCollapsePanelPress = (id: any) => {
        if (id === '3' && reposData.length === 0) {
            axios.get(props.visible.owner.repos_url).then(res => {
                setReposData(res.data);
            });
        }
    };

    return (
        <>
            {props.visible && (
                <Modal
                    width={window.innerWidth * 0.7}
                    okButtonProps={{ style: { display: 'none' } }}
                    title={props.visible.name.toUpperCase()}
                    visible={props.visible}
                    onOk={props.handleOk}
                    onCancel={props.handleCancel}
                >
                    <p>
                        <RepoIcon /> {props.visible.name}
                    </p>
                    <p>{props.visible.description}</p>
                    <div
                        style={{
                            display: 'flex',
                            flexDirection: 'row',
                            justifyContent: 'space-between',
                            maxWidth: '50%',
                            alignItems: 'center',
                        }}
                    >
                        <div style={{ display: 'flex', alignItems: 'center' }}>
                            <div
                                className='dot'
                                style={{ backgroundColor: githubLangColors[props.visible.language] }}
                            />
                            {props.visible.language}
                        </div>
                        <div>
                            <StarIcon className='icon' />
                            {props.visible.stargazers_count}
                        </div>
                        <div>
                            <GitForkIcon className='icon' />
                            {props.visible.forks}
                        </div>
                    </div>
                    <Collapse onChange={onCollapsePanelPress} style={{ marginTop: '20px' }} accordion>
                        <Panel header={props.visible.owner.type} key='1'>
                            <div>
                                <img
                                    height={'50px'}
                                    alt='avatar'
                                    className='icon'
                                    src={props.visible.owner.avatar_url}
                                />
                                {props.visible.owner.login}
                            </div>
                        </Panel>
                        <Panel header='Repo details' key='2'>
                            <p>Created: {props.visible.created_at.split('T')[0]}</p>
                            <p>Last change: {props.visible.updated_at.split('T')[0]}</p>
                            {props.visible.license && <p>License: {props.visible.license.name}</p> }
                        </Panel>
                        <Panel header={props.visible.owner.type + ' repos'} key='3'>
                            {reposData.map((data: any) => {
                                return (
                                    <>
                                        <div
                                            style={{
                                                display: 'flex',
                                                flexDirection: 'row',
                                                marginTop: '5px',
                                                alignItems: 'center',
                                            }}
                                        >
                                            <div style={{flex: 3, flexWrap: "wrap"}}>{data.name}</div>
                                            <div style={{ display: 'flex', alignItems: 'center', flex: 1 }}>
                                                <div
                                                    className='dot'
                                                    style={{ backgroundColor: githubLangColors[data.language] }}
                                                />
                                                {data.language}
                                            </div>
                                            <div style={{flex: 1}}>
                                                <StarIcon className='icon' />
                                                {data.stargazers_count}
                                            </div>
                                            <div style={{flex: 1}}>
                                                <GitForkIcon className='icon' />
                                                {data.forks}
                                            </div>
                                        </div>
                                    </>
                                );
                            })}
                        </Panel>
                    </Collapse>
                </Modal>
            )}
        </>
    );
};

export default RepoDetailsModal;
