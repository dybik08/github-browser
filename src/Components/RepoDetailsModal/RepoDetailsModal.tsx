import React, { useState } from 'react';
import { Modal } from 'antd';
import { RepoIcon } from '@primer/octicons-react';
import { Collapse } from 'antd';
import { fetchAdditionalUserRepos } from '../../actions/networkActions';
import RepoList from '../RepoList/RepoList';
import RepoInfoRow from './RepoInfoRow';
import { Repository, StringMap } from '../../Constants/types';

const { Panel } = Collapse;

const sections: StringMap = {
    '1': 'owner_type',
    '2': 'repo_details',
    '3': 'other_repos',
};

interface RepoDetailsModalProps {
    repository_data: Repository | null;
    handleOk: () => void;
    handleCancel: () => void;
}

export const RepoDetailsModal: React.FC<RepoDetailsModalProps> = ({ repository_data, handleOk, handleCancel }) => {
    const [reposData, setReposData] = useState([]);

    const onCollapsePanelPress = async (id: string | string[]) => {
        // on change function returns active tab ID, string[] if nested dropdowns
        const additionalUserReposSectionName = 'other_repos';
        if (additionalUserReposSectionName === sections[id as string] && reposData.length === 0) {
            const additionalUserRepos =
                repository_data && (await fetchAdditionalUserRepos(repository_data.owner.repos_url));
            setReposData(additionalUserRepos);
        }
    };

    return (
        repository_data && (
            <Modal
                width={window.innerWidth * 0.7}
                okButtonProps={{ style: { display: 'none' } }}
                title={repository_data.name.toUpperCase()}
                visible={repository_data !== null}
                onOk={handleOk}
                onCancel={handleCancel}
            >
                <p>
                    <RepoIcon /> {repository_data.name}
                </p>
                <p id={'repo-description'}>{repository_data.description}</p>
                <RepoInfoRow repository_data={repository_data} />
                <Collapse onChange={onCollapsePanelPress} style={{ marginTop: '20px' }} accordion>
                    <Panel header={repository_data.owner?.type} key='1'>
                        <div>
                            <img height={'50px'} alt='avatar' className='icon' src={repository_data.owner?.avatar_url} />
                            {repository_data.owner?.login}
                        </div>
                    </Panel>
                    <Panel header='Repo details' key='2'>
                        <p id='created_at'>Created: {repository_data.created_at.split('T')[0]}</p>
                        <p id='updated_at'>Last change: {repository_data.updated_at.split('T')[0]}</p>
                        {repository_data.license && <p>License: {repository_data.license?.name}</p>}
                    </Panel>
                    <Panel header={repository_data.owner?.type + ' repos'} key='3'>
                        <RepoList repositories={reposData} />
                    </Panel>
                </Collapse>
            </Modal>
        )
    );
};

export default RepoDetailsModal;
