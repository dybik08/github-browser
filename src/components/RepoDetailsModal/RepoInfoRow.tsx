import React from 'react';
import githubLangColors from '../../github-lang-colors';
import { GitForkIcon, StarIcon } from '@primer/octicons-react';
import { Repository } from '../../constants/types';

interface RepoInfoRowProps {
    repository_data: Repository;
}

export const RepoInfoRow: React.FC<RepoInfoRowProps> = ({ repository_data }) => {
    return (
        <>
            <div className='repository-details-modal-info-row'>
                <div style={{ display: 'flex', alignItems: 'center' }}>
                    <div className='dot' style={{ backgroundColor: githubLangColors[repository_data.language] }} />
                    <div id='language'>{repository_data.language}</div>
                </div>
                <div style={{ display: 'flex', alignItems: 'center' }}>
                    <StarIcon className='icon' />
                    <div id={'stargazers_count'}>{repository_data.stargazers_count}</div>
                </div>
                <div style={{ display: 'flex', alignItems: 'center' }}>
                    <GitForkIcon className='icon' />
                    <div id={'forks'}>{repository_data.forks}</div>
                </div>
            </div>
        </>
    );
};

export default RepoInfoRow;
