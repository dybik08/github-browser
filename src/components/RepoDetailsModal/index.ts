export * from './RepoDetailsModal';
export * from './RepoInfoRow';
