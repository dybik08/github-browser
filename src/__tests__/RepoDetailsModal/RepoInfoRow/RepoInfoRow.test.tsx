import React from 'react';
import { configure, shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import Adapter from 'enzyme-adapter-react-16';
import { RepoInfoRow } from '../../../Components/RepoDetailsModal/RepoInfoRow';

configure({ adapter: new Adapter() });

const mock_repository_data = {
    name: 'name',
    description: 'some description',
    language: 'javascript',
    stargazers_count: 1,
    forks: 2,
    owner: {
        repos_url: '',
        type: '',
        avatar_url: '',
        login: '',
    },
    created_at: '2020:10:12',
    updated_at: '2020:10:14',
    license: {
        name: 'MIT',
    },
};

describe('RepoInfoRow', () => {
    const wrapper: any = shallow(<RepoInfoRow repository_data={mock_repository_data} />);

    it('renders correctly', () => {
        const tree = renderer.create(wrapper).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('render correct language', () => {
        const language = wrapper.find({ id: 'language' }).text();
        expect(language).toBe('javascript');
    });

    it('render correct stargazers_count', () => {
        const stargazers_count = wrapper.find({ id: 'stargazers_count' }).text();
        expect(stargazers_count).toBe('1');
    });

    it('render correct forks', () => {
        const forks = wrapper.find({ id: 'forks' }).text();
        expect(forks).toBe('2');
    });
});
