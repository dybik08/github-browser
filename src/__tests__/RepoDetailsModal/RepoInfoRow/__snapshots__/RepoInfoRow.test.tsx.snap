// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RepoInfoRow renders correctly 1`] = `
<div
  className="repository-details-modal-info-row"
>
  <div
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
      }
    }
  >
    <div
      className="dot"
      style={
        Object {
          "backgroundColor": undefined,
        }
      }
    />
    <div
      id="language"
    >
      javascript
    </div>
  </div>
  <div
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
      }
    }
  >
    <svg
      aria-hidden="true"
      className="icon"
      dangerouslySetInnerHTML={
        Object {
          "__html": "<path fill-rule=\\"evenodd\\" d=\\"M8 .25a.75.75 0 01.673.418l1.882 3.815 4.21.612a.75.75 0 01.416 1.279l-3.046 2.97.719 4.192a.75.75 0 01-1.088.791L8 12.347l-3.766 1.98a.75.75 0 01-1.088-.79l.72-4.194L.818 6.374a.75.75 0 01.416-1.28l4.21-.611L7.327.668A.75.75 0 018 .25zm0 2.445L6.615 5.5a.75.75 0 01-.564.41l-3.097.45 2.24 2.184a.75.75 0 01.216.664l-.528 3.084 2.769-1.456a.75.75 0 01.698 0l2.77 1.456-.53-3.084a.75.75 0 01.216-.664l2.24-2.183-3.096-.45a.75.75 0 01-.564-.41L8 2.694v.001z\\"></path>",
        }
      }
      fill="currentColor"
      height={16}
      role="img"
      style={
        Object {
          "display": "inline-block",
          "userSelect": "none",
          "verticalAlign": "text-bottom",
        }
      }
      viewBox="0 0 16 16"
      width={16}
    />
    <div
      id="stargazers_count"
    >
      1
    </div>
  </div>
  <div
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
      }
    }
  >
    <svg
      aria-hidden="true"
      className="icon"
      dangerouslySetInnerHTML={
        Object {
          "__html": "<path fill-rule=\\"evenodd\\" d=\\"M12 21a1.75 1.75 0 110-3.5 1.75 1.75 0 010 3.5zm-3.25-1.75a3.25 3.25 0 106.5 0 3.25 3.25 0 00-6.5 0zm-3-12.75a1.75 1.75 0 110-3.5 1.75 1.75 0 010 3.5zM2.5 4.75a3.25 3.25 0 106.5 0 3.25 3.25 0 00-6.5 0zM18.25 6.5a1.75 1.75 0 110-3.5 1.75 1.75 0 010 3.5zM15 4.75a3.25 3.25 0 106.5 0 3.25 3.25 0 00-6.5 0z\\"></path><path fill-rule=\\"evenodd\\" d=\\"M6.5 7.75v1A2.25 2.25 0 008.75 11h6.5a2.25 2.25 0 002.25-2.25v-1H19v1a3.75 3.75 0 01-3.75 3.75h-6.5A3.75 3.75 0 015 8.75v-1h1.5z\\"></path><path fill-rule=\\"evenodd\\" d=\\"M11.25 16.25v-5h1.5v5h-1.5z\\"></path>",
        }
      }
      fill="currentColor"
      height={16}
      role="img"
      style={
        Object {
          "display": "inline-block",
          "userSelect": "none",
          "verticalAlign": "text-bottom",
        }
      }
      viewBox="0 0 24 24"
      width={16}
    />
    <div
      id="forks"
    >
      2
    </div>
  </div>
</div>
`;
